### ISSUE TITLE
Seq size is undefined after flattening

### ISSUE THREAD

**Issue by tolmasky** [facebook/immutable-js#1585]
### What happened

Seq size returns undefined after flattening.

### How to reproduce

```javascript
require("immutable").Seq([[1],2,3]).size // === 3
require("immutable").Seq([[1],2,3]).flatten().size // === undefined
require("immutable").Seq([[1],2,3]).flatten().toList().size // === 3
```

**Comment by conartist6** [comment]
This is not a bug, though I could swear the documentation for size used to be better.

Sequences are evaluated lazily, and the size property is not guaranteed to exist. As general advice I would recommend against using Seq.size for precisely that reason. I thought there used to be better documentation for the size property, but perhaps not.

The real problem is that in order to know the size of the sequence Immutable would need to evaluate it. The point of sequences is that they are evaluated lazily. In your example this means that it's actually the call to `toList` which is going to cause the flatten operation to occur. Until the flatten step is done, there's no way to know what the size of the sequence will be after flattening.

Immutable could cause access of the size property to trigger computation of the sequence, however this is probably rarely desirable behavior because sequences aren't implicitly cached. Let's say immutable did trigger evaluation of the sequence on size access and you did the following:

```
const flatSeq = require("immutable").Seq([[1],2,3]).flatten()
flat.size // 3
const flat = flatSeq.toList();
```

The above code would actually have to do the flattening twice -- once to get the size, and then again when the data is actually used. This is almost certainly not what you want.

SO / tl;dr:

You have three choices:
require("immutable").Seq([[1],2,3]).flatten().toList().size

**Comment by tolmasky** [comment]
Hi @conartist6, thank you for the detailed response. This makes total sense now.

However, I would request a different change now that I know this: It would be nice if Seq.size threw an error. It could either something like Objective-C's "Subclasser responsibility" or, ideally, something that explains this exact situation: Error("Seq is a lazy class, and can thus not be guaranteed to know the size at any give point. For this reason, you should concretize this instance (through i.e. toList()) if you need the size").

I feel that there may be some push back to do this as clearly `Seq` can *sometimes* have a size (as in your `cacheResult()` example). I am not familiar with the internals of immutable, and so I totally understand if I am off base here, but at least from a *user perspective*, I find it surprising that `Seq.size` can .... mutate. I would actually prefer it to always throw, and thus force me to take one of the two latter steps you described (either calling `.toList()` or using `iter-tools`), because as is, it makes my code potentially highly dependent on implementation details.

For example, say some function (lets call it `seqReturner`) returns a `Seq`. If it happens to call `cacheResult` internally, either explicitly, or indirectly (by, perhaps, calling `.find()` on that `Seq` at some point, which I have checked and results in the "side effect" of the receiving `Seq` knowing its size), then the returned Seq will have a `.size`.  I may then call `.size` on it and get a result back and everything is fine. If at some later point the implementor of `seqReturner` finds a way to eliminate that `.find()` call, they have in theory not broken any contract with their type signature. And yet, my calling function will now break (hopefully quite loudly, but perhaps entirely silently as the `.size`'s now `undefined` result may be interpreted as `!>= 0` in a later check for example). As such, I would have much preferred `.size` to have thrown explicitly from the beginning, forcing me to do the right thing all along. Another way to say this is that Seq really does double duty as two distinct types: `Seq` and `SeqSizeUnknown`.

Anyways, hope that all made sense, and above all, thanks for the solution!

**Comment by conartist6** [comment]
Yeah it definitely makes sense and I totally agree with you. I think it would be a great 4.0 change, since there must be people out who are actually checking for undefined and we can't break their code by throwing an error.

**Comment by conartist6** [comment]
By the way, I found the documentation I was thinking of, and the easiest method to use. It is `Collection.count`.
